From 7048a53f2abe28bae9c557ee66d90cd51040040b Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:25:47 -0400
Subject: [md] allow a partially recovered device to be hot-added to an array

Message-id: <1409563584-9000-46-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 89988
O-Subject: [RHEL7.1 PATCH 45/82] md: allow a partially recovered device to be hot-added to an array.
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: NeilBrown <neilb@suse.de>

When adding a new device into an array it is normally important to
clear any stale data from ->recovery_offset else the new device may
not be recovered properly.

However when re-adding a device which is known to be nearly in-sync,
this is not needed and can be detrimental.  The (bitmap-based)
resync will still happen, and further recovery is only needed from
where-ever it was already up to.

So if save_raid_disk is set, signifying a re-add, don't clear
->recovery_offset.

Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit 7eb418851f3278de67126ea0c427641ab4792c57)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/md.c b/drivers/md/md.c
index 26c1066..db55e48 100644
--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -7724,7 +7724,8 @@ static int remove_and_add_spares(struct mddev *mddev,
          !test_bit(Bitmap_sync, &rdev->flags)))
    continue;
 
-  rdev->recovery_offset = 0;
+  if (rdev->saved_raid_disk < 0)
+   rdev->recovery_offset = 0;
   if (mddev->pers->
       hot_add_disk(mddev, rdev) == 0) {
    if (sysfs_link_rdev(mddev, rdev))
-- 
1.7.1